### PR TITLE
pytorch metrics implementation

### PR DESCRIPTION
**Pull request by xiexinch**

### CODE REVIEW

**Comment by xvjiarui** [comment]
Fix #408

File: mmseg/core/evaluation/metrics.py [hunk 1]
```diff
@@ -49,27 +54,28 @@ def intersect_and_union(pred_label,
...
+                              ignore_index: int,
                               label_map=dict(),
                               reduce_zero_label=False):
     """Calculate Total Intersection and Union.
 
     Args:
         results (list[ndarray]): List of prediction segmentation maps.
-        gt_seg_maps (list[ndarray]): list of ground truth segmentation maps.
+        gt_seg_maps (list[ndarray]): list of ground truth segmentation
+        maps.
```
Comment (xvjiarui) [inline mmseg/core/evaluation/metrics.py:78]:
```suggestion
        gt_seg_maps (list[ndarray]): list of ground truth segmentation
               maps.
```

Comment (xvjiarui) [reply mmseg/core/evaluation/metrics.py:78]: We may update the doc string since `results` and `gt_seg_maps` could be list of str now.

File: mmseg/core/evaluation/metrics.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 import numpy as np
+import torch
 
 
 def intersect_and_union(pred_label,
                         label,
-                        num_classes,
-                        ignore_index,
+                        num_classes: int,
+                        ignore_index: int,
```
Comment (xvjiarui) [inline mmseg/core/evaluation/metrics.py:9]: We may update the docstring since `pred_label` and `label` could be str now.

File: mmseg/core/evaluation/metrics.py [hunk 3]
```diff
@@ -21,25 +22,29 @@ def intersect_and_union(pred_label,
...
+    else:
+        pred_label = torch.from_numpy((pred_label))
 
     if isinstance(label, str):
-        label = mmcv.imread(label, flag='unchanged', backend='pillow')
-    # modify if custom classes
+        label = torch.from_numpy(
+            mmcv.imread(label, flag='unchanged', backend='pillow'))
+    else:
+        label = torch.from_numpy(label)
```
Comment (xvjiarui) [inline mmseg/core/evaluation/metrics.py:42]: We may add test function for this part. 
You may create some tempfile to achieve thi.s

File: mmseg/core/evaluation/metrics.py [hunk 4]
```diff
@@ -21,25 +24,29 @@ def intersect_and_union(pred_label,
...
-         ndarray: The intersection of prediction and ground truth histogram
-             on all classes.
-         ndarray: The union of prediction and ground truth histogram on all
-             classes.
-         ndarray: The prediction histogram on all classes.
-         ndarray: The ground truth histogram on all classes.
+         torch.Tensor: The intersection of prediction and ground truth
+         histogram on all classes.
+         torch.Tensor: The union of prediction and ground truth histogram on
+         all classes.
```
Comment (xvjiarui) [inline mmseg/core/evaluation/metrics.py:30]: missing indent.

File: tests/test_metrics.py [hunk 5]
```diff
@@ -164,3 +164,44 @@ def test_mean_dice():
...
+    import cv2
+    import os
+    import shutil
+
+    def save_arr(input_arrays: list, title: str, is_image: bool):
+        filenames = []
+        PREFIX = './temp'
+        SUFFIX = '.png' if is_image else '.npy'
+        if not os.path.exists(PREFIX):
+            os.makedirs(PREFIX)
```
Comment (xvjiarui) [inline tests/test_metrics.py:179]: We may use python tempfile library https://docs.python.org/3/library/tempfile.html

**Review by xvjiarui — APPROVED** [review]

File: tests/test_metrics.py [hunk 6]
```diff
@@ -164,3 +164,42 @@ def test_mean_dice():
...
+        return filenames
+
+    pred_size = (10, 512, 1024)
+    num_classes = 19
+    ignore_index = 255
+    results = np.random.randint(0, num_classes, size=pred_size)
+    labels = np.random.randint(0, num_classes, size=pred_size)
+    labels[:, 2, 5:10] = ignore_index
+
+    temp_dir = tempfile.TemporaryDirectory()
```
Comment (xvjiarui) [inline tests/test_metrics.py:192]: `with` statement is preferred. So we won't forget to clean up. 
```
with tempfile.TemporaryDirectory() as temp_dir:
```

**Review by xvjiarui — APPROVED** [review]
